### ISSUE TITLE
SMIOL_create_decomp fails with NULL pointers

### ISSUE THREAD

**Issue by mgduda** [MiCurry/SMIOL#34]
When `NULL` pointers are provided as the actual arguments to `io_elements` and `compute_elements`, and `n_io_elements` and `n_compute_elements` are >0, `SMIOL_create_decomp` fails with a segmentation fault.

**Comment by MiCurry** [comment]
I suppose I assumed that if the user passed in `NULL` `compute_elements` and `io_elements` that they would have set `n_compute_elements` and `n_io_elements` to `0`; however, I suppose *it is* the only case in which we can check that the element arrays contain their reported amount of elements.

@mgduda, I'll update `SMIOL_create_decomp` to check to see that if `compute_elements` and `io_elements` are equal to `NULL` that `n_compute_elements` and `n_io_elements` are set to `0`. Does that seem good to you?

**Comment by mgduda** [comment]
I was thinking more of guarding against inadvertent errors when calling the `SMIOL_create_decomp` function -- if user code happened to have given the function a NULL pointer accidentally. If, say, `n_compute_elements` is > 0, but `compute_elements` is a NULL pointer, then `SMIOL_create_decomp` will fail with a segfault. In my view, it's always better to have a function fail with an error code than to have it segfault.

**Comment by mgduda** [comment]
I think simply testing that `compute_elements` and `io_elements` are non-NULL, and returning `SMIOL_INVALID_ARGUMENT` if either pointer is NULL, would be sufficient.

**Comment by mgduda** [comment]
Somewhat related to this, we get a fatal runtime error with the NAG compilers:
```
Runtime Error: smiolf.F90, line 649: Argument COMPUTE_ELEMENTS to C_LOC is a zero-sized array
Program terminated by fatal error
```
This raises the interesting question of whether we should perform additional checks in the Fortran interfaces to SMIOL routines that cannot be performed in C. For example, in Fortran, we can check on the size of an array, which is not possible in C. So, we can verify that, e.g., `size(compute_elements) >= n_compute_elements`, and we can set the `c_compute_elements` argument to `SMIOL_create_decomp` to a NULL pointer if `n_compute_elements == 0` rather than calling `c_loc`.

**Comment by mgduda** [comment]
Thinking about this more, perhaps `SMIOL_create_decomp` should only fail with `SMIOL_INVALID_ARGUMENT` if `compute_elements` is NULL *and* `n_compute_elements == 0`.

**Comment by MiCurry** [comment]
I suppose some of my confusion and the reason I wrote the current function in its current state is that I thought that `malloc` would return a NULL on `malloc(0)`. Per the POSIX man page for malloc:

> If size is 0, either a null pointer or a  unique pointer that can be successfully passed to free() shall be returned.

But, upon further investigation, `malloc` only returns a null pointer on Unix, and not Linux. So now it makes sense that upon receiving `compute_elements == NULL` or `io_elements == NULL` would be an error.

> This raises the interesting question of whether we should perform additional checks in the Fortran interfaces to SMIOL routines that cannot be performed in C. For example, in Fortran, we can check on the size of an array, which is not possible in C. So, we can verify that, e.g., size(compute_elements) >= n_compute_elements, and we can set the c_compute_elements argument to SMIOL_create_decomp to a NULL pointer if n_compute_elements == 0 rather than calling c_loc.

This makes sense. By the standard `c_loc` can only be called on a *nonzero* sized array, so we're currently violating standard. Nice that the NAG compiler picked it up.


> Thinking about this more, perhaps SMIOL_create_decomp should only fail with SMIOL_INVALID_ARGUMENT if compute_elements is NULL and n_compute_elements == 0.

Do you mean the opposite? If `compute_elements` is NULL and `n_compute_elements /= 0`? That way we can set `compute_elements` to a NULL pointer in Fortran (to avoid `c_loc`) and set `c_compute_elements` to 0. That way we can use a NULL pointer to represent an array of size 0 and return an error if there is an attribute error if `n_compute_elements /= 0`.

**Comment by MiCurry** [comment]
This issue was fixed in PR #43.
